packages/hierarchy-builder/src/hierarchy-builder/internal/operators/LabelGrouping.ts

```diff
@@ -0,0 +1,112 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import naturalCompare from "natural-compare-lite";
+import { from, mergeMap, Observable, tap, toArray } from "rxjs";
+import { Id64 } from "@itwin/core-bentley";
+import { LabelInfo } from "../../EC";
+import { HierarchyNode } from "../../HierarchyNode";
+import { getLogger } from "../../Logging";
+import { createOperatorLoggingNamespace } from "../Common";
+import { sortNodesByLabelOperator } from "./Sorting";
+
+const OPERATOR_NAME = "Grouping.ByLabel";
+/** @internal */
+export const LOGGING_NAMESPACE = createOperatorLoggingNamespace(OPERATOR_NAME);
+
+/** @internal */
+export function createLabelGroupingOperator() {
+  return function (nodes: Observable<HierarchyNode>): Observable<HierarchyNode> {
+    return nodes.pipe(
+      log((n) => `in: ${n.label}`),
+      // need all nodes in one place to group them
+      toArray(),
+      // group all nodes
+      mergeMap((resolvedNodes) => from(createLabelGroupingInformation(resolvedNodes))),
+      // convert intermediate format into a nodes observable
+      mergeMap((groupings) => {
+        const grouped = createGroupingNodes(groupings);
+        const obs = from(grouped);
+        // source observable is expected to stream sorted nodes and we're keeping them in order - only
+        // need to re-sort if we created grouping nodes
+        return grouped.hasLabelGroupingNodes ? obs.pipe(sortNodesByLabelOperator) : obs;
+      }),
+      log((n) => `out: ${n.label}`),
+    );
+  };
+}
+
+interface LabelGroupingInformation {
+  ungrouped: Array<HierarchyNode>;
+  grouped: Map<string, { labelInfo: LabelInfo; groupedNodes: Array<HierarchyNode> }>;
+}
+
+async function createLabelGroupingInformation(nodes: HierarchyNode[]): Promise<LabelGroupingInformation> {
+  const groupings: LabelGroupingInformation = { ungrouped: [], grouped: new Map() };
+  for (const node of nodes) {
+    if (HierarchyNode.isClassGroupingNode(node) && Array.isArray(node.children)) {
+      const labelGroupings = await createLabelGroupingInformation(node.children);
+      const labelGroupingNodes = createGroupingNodes(labelGroupings);
+      const sortedNodes = labelGroupingNodes.sort((lhs, rhs) => naturalCompare(lhs.label.toLocaleLowerCase(), rhs.label.toLocaleLowerCase()));
+      const newClassGroupingNode: HierarchyNode = {
+        label: node.label,
+        key: node.key,
+        children: sortedNodes,
+      };
```

Comment:
You may be losing some node's information if you don't spread it.

```suggestion
      const newClassGroupingNode: HierarchyNode = {
        ...node,
        children: sortedNodes,
      };
```

Reply:
Applied this suggestion